datamonster_api/lib/datamonster.py

```diff
@@ -199,61 +239,56 @@ def get_data(
-        """Get details (metadata) for the given datasource
+    def _avro_to_df(self, avro_buffer, split_columns):
+        """Read an avro structure into a dataframe
 
-        :param datasource_id: The ID of the datasource for which we get the details
-        :return: dictionary object with the datasource details
+        Transforms dates and columns to a stanard and agreed upon format
         """
-        path = self._get_datasource_path(datasource_id)
-        return self.client.get(path)
 
-    def _get_datasource_path(self, datasource_id):
-        return "{}/{}".format(self.datasource_path, datasource_id)
-
-    def _datasource_result_to_object(self, datasource, has_details=False):
-        ds_inst = Datasource(
-            datasource["id"],
-            datasource["name"],
-            datasource["category"],
-            datasource["uri"],
-            self,
-        )
-
-        if has_details:
-            ds_inst.set_details(datasource)
+        def parse_row(row, data_col, start_col, end_col):
+            return {
+                "value": row[data_col],
+                "start_date": pandas.to_datetime(row[start_col]),
+                "end_date": pandas.to_datetime(row[end_col]),
+                "dimensions": {
+                    split_key: row[split_key] for split_key in split_columns
+                },
+            }
+
+        reader = fastavro.reader(six.BytesIO(avro_buffer))
+        metadata = reader.schema["structure"]
+        if not metadata:
+            raise DataMonsterError(
+                "DataMonster does not currently support this request"
+            )
 
-        return ds_inst
+        if not set(metadata.keys()).issuperset(self.REQUIRED_FIELDS):
+            raise DataMonsterError(
+                "DataMonster does not currently support this request"
+            )
 
-    def _avro_to_df(self, avro_buffer):
-        """Read an avro structure into a dataframe
-        """
-        fp = six.BytesIO(avro_buffer)
-        reader = fastavro.reader(fp)
         records = [r for r in reader]
-        df = pandas.DataFrame.from_records(records)
 
-        if len(df) == 0:
-            return df
+        if not records:
+            return pandas.DataFrame.from_records(records)
 
-        # Convert date columns to datetime64 columns
-        df["upperDate"] = df["upperDate"].astype("datetime64[ns]")
-        df["lowerDate"] = df["lowerDate"].astype("datetime64[ns]")
+        start_col, end_col, data_col = [
+            metadata[col].pop() for col in self.REQUIRED_FIELDS
+        ]
+        records = [parse_row(row, data_col, start_col, end_col) for row in records]
 
-        # Create the timespan. Note we add 1 day because both dates are inclusive
-        df["time_span"] = df["upperDate"] - df["lowerDate"] + timedelta64(1, "D")
+        df = pandas.DataFrame.from_records(records)
+        df["time_span"] = df["end_date"] - df["start_date"]
+        # Change end_date to not be inclusive
```

Comment:
I think this is actually making the end_date inclusive. It comes in as exclusive. But check that math.